`CSparseTerm` & exit codes

Cryoris opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

* `CSparseTerm` structure for more flexible memory handling
* proper exit codes
* raw data access
* tests for all of the above

### Details and comments

This also includes comments done in Qiskit/qiskit#13595

Cryoris committed:
move back into 1 file

Cryoris committed:
use Arc::ptr_eq over .is()

Cryoris committed:
error codes and sparse term view

Cryoris committed:
SparseTermView -> SparesTer

Cryoris committed:
Merge branch 'py-sparse-observable' into c-sparseterms

mrossinek commented on crates/c_ext/src/exit_codes.rs:
```diff
@@ -0,0 +1,40 @@
+
+impl From<ArithmeticError> for ExitCode {
+    fn from(_value: ArithmeticError) -> Self {
+        ExitCode::ArithmeticError // do we want to cover each error enum here?
```
I am not sure I understand this comment? This would apply to the whole `impl` block and not just this one line, correct?

Cryoris replied:
Since `ArithmeticError` is an enum, we could map each enum value to a specific exit code. So 
```rust
fn from(value: ArithmeticError) -> Self {
    match value {
        ArithmeticError::MismatchedQubits => ExitCode::MismatchedQubits,
        ArithmeticError::PotentialFutureErrorThatDoesn'tExitYet => ExitCode::<...>
        ...
    }
}
```
right now just any enum value get's mapped to `ArithmeticError`. It might be nicer to have more specific error codes to look up. Given that `ArithmeticError` has only one possible value, we could maybe just add it.

mrossinek replied:
Oh I didn't know that, thanks :+1:

mrossinek commented on crates/pyext/src/lib.rs:
```diff
@@ -55,7 +55,7 @@ fn _accelerate(m: &Bound<PyModule>) -> PyResult<()> {
     add_submodule(m, ::qiskit_accelerate::sabre::sabre, "sabre")?;
     add_submodule(m, ::qiskit_accelerate::sampled_exp_val::sampled_exp_val, "sampled_exp_val")?;
-    add_submodule(m, ::qiskit_accelerate::py_sparse_observable::py_sparse_observable, "py_sparse_observable")?;
+    add_submodule(m, ::qiskit_accelerate::sparse_observable::py_sparse_observable, "py_sparse_observable")?;
```
Do we still have something called `py_sparse_observable`? I thought we got rid of that name :thinking:

Cryoris replied:
I left the module named `py_sparse_observable`, but we can also rename that. Might make more sense from a Python perspective to import `qiskit._accelerate.sparse_observable` than `qiskit._accelerate.py_sparse_observable`.... 😄

mrossinek commented on crates/c_ext/src/sparse_observable.rs:
```diff
@@ -10,130 +10,57 @@
+use qiskit_accelerate::sparse_observable::{BitTerm, SparseObservable, SparseTermView};
+use thiserror::Error;
+
+/// Errors related to C input.
+#[derive(Error, Debug)]
+pub enum CInputError {
+    #[error("Unexpected null pointer.")]
+    NullPointerError,
+    #[error("Non-aligned memory.")]
+    AlignmentError,
 }
```
Would this make more sense to have in `exit_codes` or some error-specific module?
I suspect there will be more like this in the future and then it would be weird to import this from `sparse_observable` all the time.

Cryoris replied:
Good point, I can move this 👍🏻

mrossinek commented on crates/c_ext/src/sparse_observable.rs:
```diff
@@ -10,130 +10,57 @@
-pub struct PauliTerm {
-    bit_term: BitTerm,
-    index: u32,
+pub struct CSparseTerm {
```
Would it make sense to add a comment in the docstring above pointing out that this struct is exposed under `SparseTerm` under C?
Or actually `qk_SparseTerm`? (not sure how the prefix gets applied to CamelCase names... :thinking:

Cryoris replied:
I hope we can prefix structs and enums with `Qk` and functions with `qk_` 🙈 I was wondering whether doxygen is able to do renames like this, but if not, then we should use the final names here yes.

mrossinek reviewed (commented):
Just some minor comments but overall I really like the look of this now :+1:

Cryoris committed:
review comments and other bits

- prefixes for functions
- silence some cbindgen warnings
- add obs_boundaries
- add clang-format to requirments
